mono/sgen/sgen-los.c

```diff
@@ -63,8 +65,61 @@ struct _LOSSection {
 	unsigned char *free_chunk_map;
 };
 
-/* We allow read only access on the list while sweep is not running */
-LOSObject *sgen_los_object_list = NULL;
+/* We allow read only access on the array list while sweep is not running */
+static SgenArrayList sgen_los_object_array_list = SGEN_ARRAY_LIST_INIT (NULL, sgen_array_list_default_is_slot_set, sgen_array_list_default_cas_setter, INTERNAL_MEM_PIN_QUEUE);
```

Comment:
We don't need to have `sgen_array_list_default_cas_setter`, put NULL instead. We don't have multiple threads racing for addition to the list since we only alloc large objects under the GC lock

Reply:
Yes, I can drop that. I looked through the code it seems that all read/write of next_index is done while holding GC lock. I believe I found one scenario that could end up changing the list without holding the GC lock, mono_gc_base_cleanup, but don't know if we could get activity into the list from other threads at that point in time. Looks like that code path could end up in major_finish_concurrent_collection, that will do a sgen_los_sweep, that could compact the los array and modify the next_slot variable.

Reply:
We don't care about reads, we can have concurrent reads on this structure, but not writes. Only the sweeper and the mutator write to this list. If we have the mutator running while finishing a concurrent collection, then the access to this list is the last of our worries.

Reply:
We will do a stw as part of that, so that would eliminate any mutators running, but all other places we will hold the GC lock when doing sgen_perform_collection except when called from mono_gc_base_cleanup. Any reasons why we don't hold the GC lock in that scenario as well?

Reply:
So without CAS setter I see failures in sgen-new-threads-dont-join-stw.exe during shutdown (have seen this a couple of times), but only on Linux x64 Full AOT + LLVM. Maybe this PR expose an already existing issue in this area during shutdown for this specific test. Only seen on marksweep, simple and not on marksweep-conc, simple.